atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/BucketAssignerState.java

```diff
@@ -0,0 +1,148 @@
+    }
+
+    static BucketAssignerState.WaitingClose waitingClose(long startTimestampInclusive) {
+        return ImmutableWaitingClose.of(startTimestampInclusive);
+    }
+
+    static BucketAssignerState.CloseFromOpen closeFromOpen(long startTimestampInclusive, long endTimestampExclusive) {
+        return ImmutableCloseFromOpen.of(startTimestampInclusive, endTimestampExclusive);
+    }
+
+    static BucketAssignerState.ImmediateClose immediateClose(long startTimestampInclusive, long endTimestampExclusive) {
+        return ImmutableImmediateClose.of(startTimestampInclusive, endTimestampExclusive);
+    }
+
+    @Value.Immutable
+    @JsonSerialize(as = ImmutableStart.class)
+    @JsonDeserialize(as = ImmutableStart.class)
+    interface Start extends BucketAssignerState {
+        String TYPE = "start";
+
+        @Value.Parameter
+        long startTimestampInclusive();
+
+        @Override
+        default <T> T accept(BucketAssignerState.Visitor<T> visitor) {
+            return visitor.visit(this);
+        }
+    }
+
+    @Value.Immutable
+    @JsonSerialize(as = ImmutableOpen.class)
+    @JsonDeserialize(as = ImmutableOpen.class)
+    interface Open extends BucketAssignerState {
+        String TYPE = "open";
+
+        @Value.Parameter
+        long startTimestampInclusive();
+
+        @Override
+        default <T> T accept(BucketAssignerState.Visitor<T> visitor) {
+            return visitor.visit(this);
+        }
+    }
+
+    @Value.Immutable
+    @JsonSerialize(as = ImmutableWaitingClose.class)
+    @JsonDeserialize(as = ImmutableWaitingClose.class)
+    interface WaitingClose extends BucketAssignerState {
+        String TYPE = "waitingClose";
+
+        @Value.Parameter
+        long startTimestampInclusive();
+
+        @Override
+        default <T> T accept(BucketAssignerState.Visitor<T> visitor) {
+            return visitor.visit(this);
+        }
+    }
+
+    @Value.Immutable
+    @JsonSerialize(as = ImmutableCloseFromOpen.class)
+    @JsonDeserialize(as = ImmutableCloseFromOpen.class)
+    interface CloseFromOpen extends BucketAssignerState {
+        String TYPE = "closeFromOpen";
+
+        @Value.Parameter
+        long startTimestampInclusive();
+
+        @Value.Parameter
+        long endTimestampExclusive();
+
+        @Override
+        default <T> T accept(BucketAssignerState.Visitor<T> visitor) {
+            return visitor.visit(this);
+        }
+    }
+
+    @Value.Immutable
+    @JsonSerialize(as = ImmutableImmediateClose.class)
+    @JsonDeserialize(as = ImmutableImmediateClose.class)
+    interface ImmediateClose extends BucketAssignerState {
+        String TYPE = "immediateClose";
+
+        @Value.Parameter
+        long startTimestampInclusive();
+
+        @Value.Parameter
+        long endTimestampExclusive();
+
+        @Override
+        default <T> T accept(BucketAssignerState.Visitor<T> visitor) {
+            return visitor.visit(this);
+        }
+    }
```

Comment:
do we need the bucket number to write at, or is this tracked somewhere else? I'm basically worried about something like us being in immediate close state, sleeping, and then having all records of buckets from before us getting cleaned up

Reply:
Tracked separately - I store (bucketIdentifier, state). I'm not strongly tied to storing the bucket ID in the state, but I think we may need to use <currently processing bucketIdentifier> elsewhere, and didn't want to have to parse each state to get the bucketId